### PR TITLE
Add qat

### PR DESCRIPTION
**Pull request by elad-c**
Add Quantization Aware Training model init. The new API gets a float model from the user and returns a quantized model ready for fine-tuning by the user

### CODE REVIEW

**Comment by haihabi** [comment]
Please remove the generated docs file.
We only update the docs file at a new release.

File: model_compression_toolkit/qat/keras/quantization_facade.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        tg, bit_widths_config = core_runner(in_model=in_model,
+                                            representative_data_gen=representative_data_gen,
+                                            core_config=core_config,
+                                            fw_info=fw_info,
+                                            fw_impl=fw_impl,
+                                            tpc=target_platform_capabilities,
+                                            target_kpi=target_kpi,
+                                            tb_w=tb_w)
+
+        if gptq_config is None:
```
Comment (haihabi) [inline model_compression_toolkit/qat/keras/quantization_facade.py:156]: Please remove the gptq

File: model_compression_toolkit/qat/keras/quantizer/quant_utils.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import tensorflow as tf
+from model_compression_toolkit.core.common.constants import MIN_THRESHOLD
+from typing import Tuple
+
+
+def ste_ceil(x: tf.Tensor) -> tf.Tensor:
```
Comment (haihabi) [inline model_compression_toolkit/qat/keras/quantizer/quant_utils.py:21]: This looks very similar to GPTQ maybe it should move to core?

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetric_ste.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+import tensorflow as tf
+from tensorflow_model_optimization.python.core.quantization.keras.quantize_wrapper import QuantizeWrapper
+from tensorflow.python.framework.tensor_shape import TensorShape
+from model_compression_toolkit.core.keras.quantizer.base_quantizer import BaseTrainableQuantizer
+from model_compression_toolkit.qat.keras.quantizer import quant_utils as qutils
+from model_compression_toolkit.core.common.constants import THRESHOLD
+from model_compression_toolkit.qat.common import THRESHOLD_TENSOR
+
+
+def symmetric_quantizer(input_tensor: tf.Tensor,
```
Comment (haihabi) [inline model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetric_ste.py:28]: Why not use TensorFlow fake quant ?

File: model_compression_toolkit/qat/keras/quantizer/quant_utils.py (hunk 2 above)
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/quant_utils.py:21]: it's identical, but core is definately not the place for it (IMO). Maybe a keras common for GPTQ and QAT? Or just let it be separate and diverge over time.

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetric_ste.py (hunk 3 above)
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetric_ste.py:28]: Same reason the GPTQ uses this function

Comment (haihabi) [reply model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetric_ste.py:28]: Let's talk about this on Sunday

File: model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        Instantiates a `WeightQuantizeConfig` from its config.
+
+        Args:
+            config: Output of `get_config()`.
+
+        Returns:
+            A `WeightQuantizeConfig` instance.
+        """
+
+        config['quantization_method'] = QuantizationMethod(config['quantization_method'])
```
Comment (haihabi) [inline model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py:136]: constants

File: model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    def set_quantize_weights(self, layer: Layer, quantize_weights: List[Tensor]):
+        """
+        Set the layer weights with new passed weights.
+        Args:
+            layer: Layer to set its attributes.
+            quantize_weights: Quantized weights to set as new weights.
+
+        """
+        if len(self.weight_attrs) != len(quantize_weights):
+            raise ValueError(
```
Comment (haihabi) [inline model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py:104]: Use our logger

File: model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                '`set_quantize_weights` called on layer {} with {} '
+                'weight parameters, but layer expects {} values.'.format(
+                    layer.name, len(quantize_weights), len(self.weight_attrs)))  # pragma: no cover
+
+        for weight_attr, weight in zip(self.weight_attrs, quantize_weights):
+            current_weight = getattr(layer, weight_attr)
+            if current_weight.shape != weight.shape:
+                raise ValueError('Existing layer weight shape {} is incompatible with'
```
Comment (haihabi) [inline model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py:112]: Same here

File: model_compression_toolkit/qat/pytorch/quantization_facade.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+from model_compression_toolkit.core.common import Logger
+from model_compression_toolkit.core.common.constants import PYTORCH
+from model_compression_toolkit.core.common.target_platform import TargetPlatformCapabilities
+from model_compression_toolkit.core.common.mixed_precision.kpi import KPI
+from model_compression_toolkit.core.common.framework_info import FrameworkInfo
+from model_compression_toolkit import CoreConfig
+
+
+if importlib.util.find_spec("torch") is not None:
```
Comment (haihabi) [inline model_compression_toolkit/qat/pytorch/quantization_facade.py:27]: Please take this from constants

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetric_ste.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+        return q_tensor
+
+    def get_config(self) -> Dict[str, Any]:
+        """
+        Returns: Configuration of TrainableQuantizer.
+        """
+
+        return {
+            'num_bits': self.num_bits,
```
Comment (haihabi) [inline model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetric_ste.py:144]: Use constants

File: model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        """
+        A function to get a list trainable of trainable parameters for QAT retraining from config quantizers
+
+        Returns:
+            A list of trainable Tensors
+
+        """
+        return self.weight_quantizer.get_trainable_parameters()
+
+    def get_aux_variable(self) -> List[tf.Tensor]:
```
Comment (reuvenperetz) [inline model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py:178]: Please add comments.

File: model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        return [self.weight_quantizer.get_aux_variable()]
+
+    def get_quantization_variable(self) -> List[tf.Tensor]:
```
Comment (reuvenperetz) [inline model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py:181]: Please add comments.

File: tests/keras_tests/feature_networks_tests/feature_networks/qat_test.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
+# Copyright 2021 Sony Semiconductors Israel, Inc. All rights reserved.
```
Comment (reuvenperetz) [inline tests/keras_tests/feature_networks_tests/feature_networks/qat_test.py:1]: Fix license year :)

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py (hunk 10 above)
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/configs/weight_quantizer_config.py:181]: removed the function

File: model_compression_toolkit/qat/keras/quantizer/config_factory.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    a global NoOpQuantizeConfig object.
+
+    Args:
+        n: Node to build its QuantizeConfig.
+        fw_info: Framework information (e.g., mapping from layers to their attributes to quantize).
+
+    Returns:
+        A QuantizeConfig object with the appropriate quantizers (according to the node's
+        quantization configuration).
+    """
```
[the remark this answers is not in the document]
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/config_factory.py:41]: yes....

**Pull request merged**